src/project.coffee

```diff
@@ -39,6 +39,14 @@ class Project extends Model
     @emitter = new Emitter
     @buffers ?= []
 
+    @directoryProviders = [new DefaultDirectoryProvider()]
+    atom.packages.serviceHub.consume(
+      'atom.directory-provider',
+      '^0.1.0',
+      # New providers are added to the front of @directoryProviders because
+      # DefaultDirectoryProvider is a catch-all that will always provide a Directory.
+      (provider) => @directoryProviders.splice(0, 0, provider))
```

Comment:
Do we need to be able to define alternative `Directory` implementations for local paths? Because it'd be nice to have the most common case be first in line, rather than last. We would just have to have `DefaultDirectoryProvider` return `null` if the `uri` argument had a protocol other than `file://`. That way the `Project` would be able to tell if a given path was usable or not, possibly throwing an exception like this:

```
Could not open a directory for the URI 'ftp://some-host/some/path'.
You need to install a package which implements the 'atom.directory-provider' service for that URI.
```